src/GHCMod.hs

```diff
@@ -256,6 +257,14 @@ globalArgSpec =
                reqArg "OPT" $ \g o ->
                    o { ghcUserOptions = g : ghcUserOptions o }
 
+      , option "" ["map-file"] "Redirect one file to another, --map-file \"file1.hs=file2.hs\"" $
```

Comment:
Short version:

CLI options:
* `--map-file "file1.hs=file2.hs"` can be used to tell ghc-mod that it should take source code for `file1.hs` from `file2.hs`. `file1.hs` can be either full path, or path relative to project root. `file2.hs` has to be either relative to current directory, or full path (preferred).
* `--map-file "file.hs"` can be used to tell ghc-mod that it should take source code for `file.hs` from stdin. File end marker is `\EOT\n`, i.e. `\x04\x0A`. `file.hs` may or may not exist, and should be either full path, or relative to project root.

Interactive commands:
* `map-file file.hs` -- tells ghc-modi to read `file.hs` source from stdin. Works the same as second form of `--map-file` CLI command.
* `unmap-file file.hs` -- unloads previously mapped file, so that it's no longer mapped. `file.hs` can be full path or relative to project root, either will work.

Exposed functions (this probably needs more thought, since exported functions are somewhat low-level):
* `getMMappedFiles` -- returns a `Map FilePath FileMapping` in `GmState` monad. This map is used for file name/source substitutions.
* `setMMappedFiles` -- allows to set said map.

Should probably export `loadMappedFile` and `unloadMappedFile` instead.